### PR TITLE
CharReader: Add StructuredError

### PR DESCRIPTION
**Pull request by martinduffy1**
The move from Json::Reader to Json::CharReader omitted the ability to get structured error messages from the parser. Clients which want to extract the line and column information have to parse the returned string. Add a new getStructuredErrors() method which exposes the error as structured data instead of a string.

Based on the work from @kylefromkitware and comments from @BillyDonahue on #1281.

### CODE REVIEW

**Comment by ahedberg** [comment]
Out of curiosity, what's the status of this PR? We'd find this addition useful as well.

File: src/lib_json/json_reader.cpp [hunk 1]
```diff
@@ -1976,6 +1974,15 @@ void CharReaderBuilder::setDefaults(Json::Value* settings) {
   //! [CharReaderBuilderDefaults]
 }
 
+auto CharReader::getStructuredErrors() const -> std::vector<CharReader::StructuredError> {
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1977]:
```suggestion
std::vector<CharReader::StructuredError> CharReader::getStructuredErrors() const {
```

File: src/lib_json/json_reader.cpp [hunk 2]
```diff
@@ -1976,6 +1974,15 @@ void CharReaderBuilder::setDefaults(Json::Value* settings) {
...
+    return _impl->getStructuredErrors();
+}
+
+bool CharReader::parse(char const* beginDoc, char const* endDoc, Value* root,
+                     String* errs) {
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1982]:
```suggestion
bool CharReader::parse(char const* beginDoc, char const* endDoc, Value* root,
                       String* errs) {
```

**Comment by BillyDonahue** [comment]
> Out of curiosity, what's the status of this PR? We'd find this addition useful as well.

I have some pending comments that I never hit "Send" on. Sorry about that.

**Review by BillyDonahue — CHANGES_REQUESTED** [review]

File: include/json/reader.h [hunk 3]
```diff
@@ -337,7 +363,7 @@ class JSON_API CharReaderBuilder : public CharReader::Factory {
   CharReaderBuilder();
   ~CharReaderBuilder() override;
 
-  CharReader* newCharReader() const override;
+  CharReader::Impl* newCharReader() const override;
```
[the remark this answers is not in the document]
Comment (martinduffy1) [reply include/json/reader.h:366]: I think I'm not understanding your suggestion about this here (https://github.com/open-source-parsers/jsoncpp/pull/1281#issuecomment-816927120)

> Close but I would not do `dynamic_cast` of `this`, but rather hold a polymorphic `_impl`. `OurCharReader` would implement a polymorphic interface and implement `getStructuredErrors()` directly.
> 
> So it would be:
> 
> ```c++
> CharReader {
> public:
>     std::vector<StructuredError> getStructuredErrors() const;
> 
>     /** apologetically virtual only for ABI compatibility. Not an extension point, so `final`. */
>     virtual bool parse(etc...) final;
> 
> private:
>     class Impl;  // Privately provides the polymorphic interface that CharReader provides today.
>     std::unique_ptr<Impl> _impl;
> };
> 
> // In cpp file to prevent inlining and control visibility.
> 
> class CharReader::Impl {
> public:
>     virtual ~Impl() = default;
>     virtual bool parse(etc...) = 0;
>     virtual std::vector<StructuredError> getStructuredErrors() const = 0;
> };
> 
> auto CharReader::getStructuredErrors() const -> std::vector<StructuredError> {
>     return _impl->getStructuredErrors();
> }
> 
> bool CharReader::parse(etc...) {
>     return _impl->parse(etc...);
> }
> ```

This particular change was part of a series of changes related to making `OurCharReader` override `CharReader::Impl`. I did this since you suggested `OurCharReader` implement `getStructuredErrors()` while `CharReader::getStructuredErrors()` calls `_impl->getStructuredErrors()`. Can you explain how you envisioned `_impl` getting the implementation of `getStructuredErrors()` in your suggestion for this?

[the remark this answers is not in the document]
Comment (martinduffy1) [reply include/json/reader.h:366]: @BillyDonahue can you get back to me about this question so I can finish resolving your comments?

[the remark this answers is not in the document]
Comment (BillyDonahue) [reply include/json/reader.h:366]: Yes sorry for the delay.

I wasn't sure what the question meant by "override" of `CharReader::Impl`, so I was waiting until I had a good amount of time to spend on it. Okay.

I'm thinking that `newCharReader` just returns `CharReader*` as it always has.

ABI-wise, no client code depends on the internals of `CharReader`.  They only ever work with it by pointer and by calling functions on those pointers that live in the JsonCpp library.

But the `newCharReader` function can place whatever kind of `Impl` it wants into the `CharReader` it returns. This hides the details from the callers.

It would make an `Impl` subclass that has a `getStructuredErrors` appropriate for a `CharReader`, and place that into the `CharReader` to which it returns a pointer.

**Comment by martinduffy1** [comment]
@BillyDonahue I responded to all the comments from your last review. Let me know if you have any questions.

File: src/lib_json/json_reader.cpp [hunk 4]
```diff
@@ -1873,20 +1868,36 @@ std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
...
-public:
   OurCharReader(bool collectComments, OurFeatures const& features)
-      : collectComments_(collectComments), reader_(features) {}
-  bool parse(char const* beginDoc, char const* endDoc, Value* root,
-             String* errs) override {
-    bool ok = reader_.parse(beginDoc, endDoc, *root, collectComments_);
-    if (errs) {
-      *errs = reader_.getFormattedErrorMessages();
+      : collectComments_(collectComments), reader_(features) {
+        this->_impl = std::unique_ptr<OurCharReader::Impl>(new OurCharReader::Impl(collectComments, features));
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1877]: If you're in the OurCharReader class scope you don't need to say OurCharReader here.
You can initialize _impl just like you initialized the other members, in the initialization list.

File: include/json/reader.h [hunk 5]
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
...
+                     String* errs) final;
+
+   /** \brief Returns a vector of structured errors encountered while parsing.
+   *
+   * \return A (possibly empty) vector of StructuredError objects.
+   */
+  std::vector<StructuredError> getStructuredErrors() const;
+
+  class Impl {
+    public:
```
Comment (BillyDonahue) [inline include/json/reader.h:282]: Use surrounding indentation style.

File: include/json/reader.h [hunk 6]
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
    * error occurred.
    */
   virtual bool parse(char const* beginDoc, char const* endDoc, Value* root,
-                     String* errs) = 0;
+                     String* errs) final;
+
+   /** \brief Returns a vector of structured errors encountered while parsing.
```
Comment (BillyDonahue) [inline include/json/reader.h:275]: This doc should answer some questions about the way CharReader stores the errors after a parse.

Does each parse call reset the stored errors vector?
Does a call to this function consume and reset the errors?

File: include/json/reader.h [hunk 7]
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
    * error occurred.
    */
   virtual bool parse(char const* beginDoc, char const* endDoc, Value* root,
-                     String* errs) = 0;
+                     String* errs) final;
```
Comment (BillyDonahue) [inline include/json/reader.h:273]: remove `final`

**Review by BillyDonahue — CHANGES_REQUESTED** [review]

File: src/lib_json/json_reader.cpp [hunk 8]
```diff
@@ -1873,20 +1868,36 @@ std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
...
+      bool ok = reader_.parse(beginDoc, endDoc, *root, collectComments_);
+      if (errs) {
+        *errs = reader_.getFormattedErrorMessages();
+      }
+      return ok;
     }
-    return ok;
-  }
+
+    std::vector<CharReader::StructuredError> getStructuredErrors() const {
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1897]: add `override`

File: src/lib_json/json_reader.cpp [hunk 9]
```diff
@@ -1873,20 +1868,36 @@ std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
...
-    bool ok = reader_.parse(beginDoc, endDoc, *root, collectComments_);
-    if (errs) {
-      *errs = reader_.getFormattedErrorMessages();
+      : collectComments_(collectComments), reader_(features) {
+        this->_impl = std::unique_ptr<OurCharReader::Impl>(new OurCharReader::Impl(collectComments, features));
+      }
+
+  class Impl : public CharReader::Impl {
+  public:
+    bool const collectComments_;
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1882]: I think the convention here is for data members to be placed at the bottom of the class def

**Review by BillyDonahue — CHANGES_REQUESTED** [review]

File: include/json/reader.h [hunk 10]
```diff
@@ -262,16 +270,34 @@ class JSON_API CharReader {
...
     virtual ~Factory() = default;
-    /** \brief Allocate a CharReader via operator new().
+    /** \brief Allocate a CharReader Impl via operator new().
      * \throw std::exception if something goes wrong (e.g. invalid settings)
      */
     virtual CharReader* newCharReader() const = 0;
   }; // Factory
+
+protected:
+  std::unique_ptr<Impl> _impl;
```
[the remark this answers is not in the document]
Comment (martinduffy1) [reply include/json/reader.h:299]: It needs to be protected because it is set from OurCharReader

File: src/lib_json/json_reader.cpp (hunk 4 above)
Comment (martinduffy1) [reply src/lib_json/json_reader.cpp:1877]: You cannot initialize base class members from the initialization list.

**Comment by martinduffy1** [comment]
Thank you for the quick review @BillyDonahue 
I added some comments about why I am handling impl_ the way I am, and fixed everything else according to your suggestions.

File: src/lib_json/json_reader.cpp [hunk 11]
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
...
   OurCharReader(bool collectComments, OurFeatures const& features)
-      : collectComments_(collectComments), reader_(features) {}
-  bool parse(char const* beginDoc, char const* endDoc, Value* root,
-             String* errs) override {
-    bool ok = reader_.parse(beginDoc, endDoc, *root, collectComments_);
-    if (errs) {
-      *errs = reader_.getFormattedErrorMessages();
+      : collectComments_(collectComments), reader_(features) {
+        this->_impl=std::unique_ptr<Impl>(new Impl(collectComments, features));
+      }
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1878]: fix indent

File: src/lib_json/json_reader.cpp [hunk 12]
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
...
 public:
   OurCharReader(bool collectComments, OurFeatures const& features)
-      : collectComments_(collectComments), reader_(features) {}
-  bool parse(char const* beginDoc, char const* endDoc, Value* root,
-             String* errs) override {
-    bool ok = reader_.parse(beginDoc, endDoc, *root, collectComments_);
-    if (errs) {
-      *errs = reader_.getFormattedErrorMessages();
+      : collectComments_(collectComments), reader_(features) {
+        this->_impl=std::unique_ptr<Impl>(new Impl(collectComments, features));
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1877]: Spaces around the `=` here.
But I think this assignment can be eliminated by calling a protected CharReader constructor.

File: include/json/reader.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (BillyDonahue) [reply include/json/reader.h:298]: `CharReader` could have a protected constructor that accepts a `std::unique_ptr<Impl>` and moves it into the `_impl` data member. `OurCharReader` could call that constructor.

File: src/lib_json/json_reader.cpp (hunk 12 above)
Comment (BillyDonahue) [reply src/lib_json/json_reader.cpp:1877]: (line 1871) If the `impl` is recording the `collectComments`, then `OurCharReader` shouldn't have a redundant copy of it.  I think `collectComments_` and `reader_` can be removed from the `OurCharReader` class.

File: src/lib_json/json_reader.cpp [hunk 13]
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
...
+      }
+
+  class Impl : public CharReader::Impl {
+    bool const collectComments_;
+    OurReader reader_;
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1882]: On Line 1872, `OurCharReader` has an `OurReader` data member.
Isn't this `reader_` redundant with that one?

File: include/json/reader.h [hunk 14]
```diff
@@ -262,7 +268,22 @@ class JSON_API CharReader {
...
+                     String* errs);
+
+  /** \brief Returns a vector of structured errors encountered while parsing.
+   * Each parse call resets the stored list of errors. Each call to this
+   * function reads the stored list of errors and builds a vector of
+   * structured errors to return.
+  */
+  std::vector<StructuredError> getStructuredErrors() const;
+
+  class Impl {
```
Comment (BillyDonahue) [inline include/json/reader.h:280]: I think this should be `protected` API, not `public`.

Derived classes need to know what an `Impl` is but nobody else does.

**Review by BillyDonahue — CHANGES_REQUESTED** [review]

**Comment by martinduffy1** [comment]
@BillyDonahue thanks for your review, I've addressed these comments.

**Comment by BillyDonahue** [comment]
> @BillyDonahue thanks for your review, I've addressed these comments.

I still see some comments unaddressed.

**Comment by martinduffy1** [comment]
Which comments are unaddressed? Is the problem that I didn't click resolve
on the comments?
I did notice another indentation error though which I will commit now.

File: src/lib_json/json_reader.cpp (hunk 13 above)
Comment (BillyDonahue) [reply src/lib_json/json_reader.cpp:1879]: not addressed, for example.

Comment (martinduffy1) [reply src/lib_json/json_reader.cpp:1879]: I addressed this by removing the OurReader data member from OurCharReader

File: src/lib_json/json_reader.cpp [hunk 15]
```diff
@@ -1878,15 +1873,29 @@ class OurCharReader : public CharReader {
...
+
+  public:
+    Impl(bool collectComments, OurFeatures const& features)
+        : collectComments_(collectComments), reader_(features) {}
+    bool parse(char const* beginDoc, char const* endDoc, Value* root,
```
[the remark this answers is not in the document]
Comment (martinduffy1) [reply src/lib_json/json_reader.cpp:1885]: This isn't showing as Outdated, but I added blank lines between these.

**Comment by martinduffy1** [comment]
> > Which comments are unaddressed? Is the problem that I didn't click resolve on the comments? I did notice another indentation error though which I will commit now.
> 
> Please look through some of those comments. I think you'll see what I mean.

The comment you responded to was addressed elsewhere in the code. I added some comments to the issues not showing as Outdated to explain how these were addressed. In the future, I'll make sure to include a description of how the comments were addressed to make it more straightforward to review.

**Comment by martinduffy1** [comment]
> > > Which comments are unaddressed? Is the problem that I didn't click resolve on the comments? I did notice another indentation error though which I will commit now.
> > 
> > 
> > Please look through some of those comments. I think you'll see what I mean.
> 
> The comment you responded to was addressed elsewhere in the code. I added some comments to the issues not showing as Outdated to explain how these were addressed. In the future, I'll make sure to include a description of how the comments were addressed to make it more straightforward to review.

@BillyDonahue can you take a look at this comment and let me know if there are still changes you are waiting on from me, or if you just haven't gotten a chance to review them yet?

**Comment by BillyDonahue** [comment]
> > > > Which comments are unaddressed? Is the problem that I didn't click resolve on the comments? I did notice another indentation error though which I will commit now.
> > > 
> > > 
> > > Please look through some of those comments. I think you'll see what I mean.
> > 
> > 
> > The comment you responded to was addressed elsewhere in the code. I added some comments to the issues not showing as Outdated to explain how these were addressed. In the future, I'll make sure to include a description of how the comments were addressed to make it more straightforward to review.
> 
> @BillyDonahue can you take a look at this comment and let me know if there are still changes you are waiting on from me, or if you just haven't gotten a chance to review them yet?

Sorry it's very hard to follow Github PR reviews, so if something is addressed elsewhere I may not make the connection. I will take another deeper look.

File: src/lib_json/json_reader.cpp (hunk 9 above)
Comment (BillyDonahue) [reply src/lib_json/json_reader.cpp:1878]: Ok, this one wasn't done. Can you please move these data members to the bottom of the class, after the member functions?

File: src/lib_json/json_reader.cpp [hunk 16]
```diff
@@ -1860,10 +1855,10 @@ String OurReader::getFormattedErrorMessages() const {
   return formattedMessage;
 }
 
-std::vector<OurReader::StructuredError> OurReader::getStructuredErrors() const {
-  std::vector<OurReader::StructuredError> allErrors;
+std::vector<CharReader::StructuredError> OurReader::getStructuredErrors() const {
```
Comment (BillyDonahue) [inline src/lib_json/json_reader.cpp:1858]: The api before this change had an OurReader::getStructuredErrors that returned a std::vector<OurReader::StructuredError>. But now the return type has changed its name, which is an API change, which was I thought what we were trying to avoid in the original proposal by not changing the offsets of the virtual functions in the CharReader class.

Comment (martinduffy1) [reply src/lib_json/json_reader.cpp:1858]: I've moved those data members and made them private.

I resolved the api change by adding a transformation in OurImpl::getStructuredErrors to convert the CharReader vector to an OurReader vector.

Comment (martinduffy1) [reply src/lib_json/json_reader.cpp]: @BillyDonahue can you let me know when you get a chance to take a look at this?

**Review by BillyDonahue — CHANGES_REQUESTED** [review]

**Comment by martinduffy1** [comment]
@BillyDonahue let me know if these changes address your comment

**Comment by martinduffy1** [comment]
@BillyDonahue it's been 8 weeks since your last update on this. Is there anything you're waiting on from me?

**Comment by BillyDonahue** [comment]
> @BillyDonahue it's been 8 weeks since your last update on this. Is there anything you're waiting on from me?

Sorry. I made a few suggestions and didn't hear back. If you have resolved them can you comment and resolve?

**Review by BillyDonahue — APPROVED** [review]

**Comment by martinduffy1** [comment]
@BillyDonahue I see this marked as approved. Is there a release timeline that I can expect for these changes to get merged?

**Comment by BillyDonahue** [comment]
> @BillyDonahue I see this marked as approved. Is there a release timeline that I can expect for these changes to get merged?

Sorry I thought the approval was sufficient for you to merge it rather than me.
I see that we have to rerun Travis.

**Comment by BillyDonahue** [comment]
Hm. I don't actually know how to restart the stuck Travis CI job.
I poked around and didn't find anything.
If you push an update maybe it will kickstart itself.

**Comment by martinduffy1** [comment]
@BillyDonahue I squashed the commits and pushed to try to retrigger the CI job. I don't think I can merge this as I don't have write access to the repository. I'm also wondering if there can be a minor version increase when this and #1412 are merged so that we can detect versions with these updates.
